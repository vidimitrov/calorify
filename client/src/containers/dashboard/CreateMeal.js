import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from '@reach/router';
import { isEmpty } from 'lodash';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import ArrowBack from '@material-ui/icons/ArrowBack';
import Button from '@material-ui/core/Button';
import Snackbar from '@material-ui/core/Snackbar';
import CustomSnackbar from '../../components/common/CustomSnackbar';
import Form from '../../components/dashboard/Form';
import FormControls from '../../components/dashboard/FormControls';
import InputWrapper from '../../components/dashboard/InputWrapper';
import Logo from '../../components/dashboard/Logo';
import Wrapper from '../../components/dashboard/Wrapper';
import CustomTypography from '../../components/dashboard/CustomTypography';
import CustomIconButton from '../../components/dashboard/CustomIconButton';
import CustomTextField from '../../components/dashboard/CustomTextField';
import ValidationLabel from '../../components/dashboard/ValidationLabel';
import logo from '../../assets/img/logo.png';
import { createMeal as createMealActionCreator } from '../../actions/meals';
import { convertDateToUTC } from '../../lib/dates';

export class CreateMeal extends React.Component {
  constructor(props) {
    super(props);

    const now = props.initialDate || new Date();
    const dateWithTimeZone = convertDateToUTC(now);
    const splittedString = dateWithTimeZone.toISOString().split(':');
    const defaultDate = `${splittedString[0]}:${splittedString[1]}`;

    this.state = {
      name: '',
      calories: '',
      date: defaultDate,
      requiredName: false,
      requiredCalories: false,
      change: false,
      positiveSnackbarOpen: false,
      negativeSnackbarOpen: false,
    };

    this.onChangeHandler = this.onChangeHandler.bind(this);
    this.handlePositiveSnackbarClose = this.handlePositiveSnackbarClose.bind(this);
    this.handleNegativeSnackbarClose = this.handleNegativeSnackbarClose.bind(this);
  }

  onChangeHandler(key, value) {
    let { requiredName, requiredCalories } = this.state;

    if (key === 'name') {
      requiredName = false;
    }

    if (key === 'calories') {
      requiredCalories = false;
    }

    this.setState({
      change: true,
      requiredName,
      requiredCalories,
      [key]: value,
    });
  }

  handlePositiveSnackbarClose(event, reason) {
    if (reason === 'clickaway') {
      return;
    }

    this.setState({ positiveSnackbarOpen: false });
  }

  handleNegativeSnackbarClose(event, reason) {
    if (reason === 'clickaway') {
      return;
    }

    this.setState({ negativeSnackbarOpen: false });
  }

  render() {
    const {
      user,
      createMeal,
    } = this.props;
    const {
      name,
      calories,
      date,
      requiredName,
      requiredCalories,
      change,
      positiveSnackbarOpen,
      negativeSnackbarOpen,
    } = this.state;
    const { navigate } = this.props;

    if (!user) {
      return (
        <Redirect to="/auth/login" />
      );
    }

    return (
      <Wrapper data-testid="create-meal-page">
        <AppBar position="static">
          <Toolbar>
            <CustomIconButton color="secondary" aria-label="Menu">
              <ArrowBack onClick={() => navigate('/')} />
            </CustomIconButton>
            <Logo src={logo} alt="logo" />
            <CustomTypography variant="h6" color="secondary">
              Create Meal
            </CustomTypography>
          </Toolbar>
        </AppBar>
        <Form>
          <FormControls>
            <InputWrapper>
              <CustomTextField
                placeholder="Meal name"
                value={name}
                name="name"
                error={requiredName}
                onChange={e => this.onChangeHandler('name', e.target.value)}
              />
              <ValidationLabel
                name={requiredName ? 'required-name' : ''}
                invalid={requiredName}
              >
                Meal name is required
              </ValidationLabel>
            </InputWrapper>
            <InputWrapper>
              <CustomTextField
                placeholder="Number of calories"
                value={calories}
                name="calories"
                error={requiredCalories}
                onChange={e => this.onChangeHandler('calories', e.target.value)}
              />
              <ValidationLabel
                name={requiredCalories ? 'required-calories' : ''}
                invalid={requiredCalories}
              >
                Meal calories are required
              </ValidationLabel>
            </InputWrapper>
            <InputWrapper>
              <CustomTextField
                placeholder="Date (optional)"
                value={date}
                type="datetime-local"
                InputLabelProps={{
                  shrink: true,
                }}
                onChange={e => this.onChangeHandler('date', e.target.value)}
              />
            </InputWrapper>
          </FormControls>
          {/* TODO: It should be disabled if no changes are made */}
          <Button
            variant="contained"
            color="primary"
            name="submit-create-meal"
            disabled={!change}
            onClick={() => {
              if (!name || !calories) {
                this.setState({
                  requiredName: !name,
                  requiredCalories: !calories,
                });
              } else {
                const isoDate = convertDateToUTC(new Date(date)).toISOString();
                const formattedDate = isoDate.split('T')[0];
                const formattedTime = isoDate.split('T')[1].split('Z')[0];

                createMeal(name, calories, formattedDate, formattedTime)
                  .then(() => {
                    this.setState({
                      positiveSnackbarOpen: true,
                    });
                    setTimeout(() => {
                      navigate('/');
                    }, 1000);
                  }).catch(() => {
                    this.setState({
                      negativeSnackbarOpen: true,
                    });
                  });
              }
            }}
          >
            Create Meal
          </Button>
        </Form>
        <Snackbar
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'left',
          }}
          open={positiveSnackbarOpen}
          autoHideDuration={1000}
          onClose={this.handlePositiveSnackbarClose}
        >
          <CustomSnackbar
            onClose={this.handlePositiveSnackbarClose}
            variant="success"
            message="Meal created successfully"
          />
        </Snackbar>
        <Snackbar
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'left',
          }}
          open={negativeSnackbarOpen}
          autoHideDuration={1000}
          onClose={this.handleNegativeSnackbarClose}
        >
          <CustomSnackbar
            onClose={this.handleNegativeSnackbarClose}
            variant="error"
            message="Couldn\'t create meal. Please try again"
          />
        </Snackbar>
      </Wrapper>
    );
  }
}

CreateMeal.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.string.isRequired,
    role: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired,
    dailyCaloriesLimit: PropTypes.number.isRequired,
  }),
  createMeal: PropTypes.func.isRequired,
  navigate: PropTypes.func.isRequired,
  initialDate: PropTypes.instanceOf(Date),
};

CreateMeal.defaultProps = {
  user: null,
  initialDate: null,
};

const mapStateToProps = state => ({
  user: isEmpty(state.user.data) ? null : state.user.data,
});

const mapDispatchToProps = dispatch => ({
  createMeal: (name, calories, date, time) => dispatch(
    createMealActionCreator({
      name,
      calories,
      date,
      time,
    }),
  ),
});

export default connect(mapStateToProps, mapDispatchToProps)(CreateMeal);
