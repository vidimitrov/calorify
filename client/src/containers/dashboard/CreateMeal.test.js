import React from 'react';
import renderer from 'react-test-renderer';
import { CreateMeal } from './CreateMeal';

it('CreateMeal container renders correctly', () => {
  const props = {
    createMeal: () => { },
    navigate: () => { },
    account: {
      id: 'some-fake-id',
      name: 'John',
      dailyCaloriesLimit: 2000,
      role: 'user',
      email: 'john@email.com',
    },
    initialDate: new Date('1995-12-17T03:24:00'),
  };
  const tree = renderer.create(
    <CreateMeal {...props} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
