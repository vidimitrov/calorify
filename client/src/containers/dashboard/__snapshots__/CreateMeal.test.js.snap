// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateMeal container renders correctly 1`] = `
<div
  className="Wrapper-m60uq5-0 hkWJUG"
  data-testid="create-meal-page"
>
  <header
    className="MuiPaper-root-10 MuiPaper-elevation4-16 MuiAppBar-root-1 MuiAppBar-positionStatic-5 MuiAppBar-colorPrimary-8"
  >
    <div
      className="MuiToolbar-root-37 MuiToolbar-regular-39 MuiToolbar-gutters-38"
    >
      <button
        aria-label="Menu"
        className="MuiButtonBase-root-47 MuiIconButton-root-41 MuiIconButton-colorSecondary-44 CustomIconButton-sc-1edavx9-0 eiNqgO"
        disabled={false}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-46"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-50"
            focusable="false"
            onClick={[Function]}
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-154"
        />
      </button>
      <img
        alt="logo"
        className="Logo-sc-1vcm3wd-0 lmrySJ"
        src="logo.png"
      />
      <h6
        className="MuiTypography-root-59 MuiTypography-h6-76 MuiTypography-colorSecondary-90 CustomTypography-e31nmz-0 kgGAHm"
      >
        Create Meal
      </h6>
    </div>
  </header>
  <div
    className="Form-ycfidt-0 jwaFmv"
  >
    <div
      className="FormControls-sc-7fqkjw-0 dcNCfX"
    >
      <div
        className="InputWrapper-sc-1a0h3fv-0 hchKYo"
      >
        <div
          className="MuiFormControl-root-94 CustomTextField-uzoay-0 ewXInx"
        >
          <div
            className="MuiInputBase-root-111 MuiInput-root-98 MuiInput-underline-102 MuiInputBase-formControl-112 MuiInput-formControl-99"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-121 MuiInput-input-106"
              disabled={false}
              name="name"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Meal name"
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
        <p
          className="ValidationLabel-sc-1uikjc9-0 gbgFrP"
          name={false}
        >
          Meal name is required
        </p>
      </div>
      <div
        className="InputWrapper-sc-1a0h3fv-0 hchKYo"
      >
        <div
          className="MuiFormControl-root-94 CustomTextField-uzoay-0 ewXInx"
        >
          <div
            className="MuiInputBase-root-111 MuiInput-root-98 MuiInput-underline-102 MuiInputBase-formControl-112 MuiInput-formControl-99"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-121 MuiInput-input-106"
              disabled={false}
              name="calories"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Number of calories"
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
        <p
          className="ValidationLabel-sc-1uikjc9-0 gbgFrP"
          name={false}
        >
          Meal calories are required
        </p>
      </div>
      <div
        className="InputWrapper-sc-1a0h3fv-0 hchKYo"
      >
        <div
          className="MuiFormControl-root-94 CustomTextField-uzoay-0 ewXInx"
        >
          <div
            className="MuiInputBase-root-111 MuiInput-root-98 MuiInput-underline-102 MuiInputBase-formControl-112 MuiInput-formControl-99"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              className="MuiInputBase-input-121 MuiInput-input-106 MuiInputBase-inputType-124 MuiInput-inputType-109"
              disabled={false}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Date (optional)"
              required={false}
              type="datetime-local"
              value="1995-12-17T03:24"
            />
          </div>
        </div>
      </div>
    </div>
    <button
      className="MuiButtonBase-root-47 MuiButtonBase-disabled-48 MuiButton-root-128 MuiButton-contained-139 MuiButton-containedPrimary-140 MuiButton-raised-142 MuiButton-raisedPrimary-143 MuiButton-disabled-148"
      disabled={true}
      name="submit-create-meal"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="button"
    >
      <span
        className="MuiButton-label-129"
      >
        Create Meal
      </span>
    </button>
  </div>
</div>
`;
